### ISSUE TITLE
hash encoding coding snippet

### ISSUE THREAD

**Issue by pure-water** [NVlabs/instant-ngp#112]
Hi Is there a chance to give a quick note on where the  meat is regarding to the "hash" part?  I looked into the code, the nerf in TCNN seems to be a vanilla Nerf? 

Regards
Yao

**Comment by Tom94** [comment]
Hi there, the grid encoding is part of the __tiny-cuda-nn__ framework; you can find it here: https://github.com/NVlabs/tiny-cuda-nn/blob/master/include/tiny-cuda-nn/encodings/grid.h

It can be configured with dense, tiled, and hash-based memory, the latter being the encoding described in the paper.

**Comment by pure-water** [comment]
Hi thanks, actually find it already in the stand-alone TCNN repo.

**Comment by pure-water** [comment]
This "hash" seems to be compressing the traditional spatial encoding a bit. This is quite un-intution but apparently it works!  What a magic!

**Comment by Tom94** [comment]
They are a replacement to `x, y, z`. Theta/phi (or another parameterization of direction) are fed separately in the NeRF setting (controlled by the `dir_encoding` entry of the network config .json).

**Comment by pure-water** [comment]
Sorry, my bad wording. I am aware that is a bit of specific of nerf details whereby the color and density are seperated. Let me reclariy my question on this context, let us focus on density only first( no noise/confusion of the theta/phi anyway). 　I mean from the paper the hash-grid-encoding output are cascaded with <x,y,z>. You are saying it is a replacement of <x,y,z>? (Figure 3, <c>)

Regards
Yao

**Comment by Tom94** [comment]
The encoding is not concatenated (do you mean this with "cascaded" or something else?) with <x,y,z>. It really is used as a replacement to inputing <x,y,z> to the network (plus optional extra dimensions \xi, which, to be fair, could include <x,y,z> if you'd like. But we don't do that in any of the paper results.

Could you point me to the place in the paper that suggests we concatenate with <x,y,z>? We may need to clarify that.

**Comment by pure-water** [comment]
Regarding to the Nerf case (It seems  I can not attch image on this forum?). 

It is here [https://nvlabs.github.io/instant-ngp/assets/mueller2022instant.pdf](url)

Page 4, Figure 3 , part (c). It appears that "E" here actually means "x,y,z" in the Nerf Context? 

Regards
Yao

**Comment by pure-water** [comment]
Thanks. If we simply focus on the density network, yup, What Thomas said is aligned with the code as pointed out by resurgo97.  So I think the "auxilary input" in the paper doesn't apply for the Nerf-Case?

Regards
Yao

**Comment by resurgo97** [comment]
Right. I also think the paper could be a little more clear on that.

**Comment by mmalex** [comment]
Hi! thanks for the feedback about clarity in the paper! You are right, in
the NeRF case we feed the output of the hashgrid encoding directly to a
first MLP without any additional 'auxiliary inputs' to predict a density
that is not view-dependent, and then 'chain' the first 16 neurons of hidden
state from that MLP along with an (spherical harmonic or one blob) encoded
view direction into the 'rgb network' to predict color. We also implemented
nerf previously using the simpler approach of concatenating the view
direction as an aux input alongside the hashgrid encoded position, and
using a single MLP; this also works fine, but leads to a view dependent
density. When we came to make comparisons to other nerf techniques, we
decided instead to use the more complicated setup since it is closer to
other nerfs. But both methods work fine, and the hashgrid encoding itself
can easily be used with or without auxiliary inputs, depending on the needs
of the particular application.

**Comment by pure-water** [comment]
All right. Thanks mmalex for clear clarification which is helpful. 
Essentially just as a hypothetical example, if we do not care about the density at all,only the final <R,G,B> color ( I understand the color is depended on density in principle).   But in the applicaton case, if we only need <R,G,B>  we should be able to put <x,y,z,thea,phi> together as the input of the hash-grid and connect it to the rgb-network? 

Regards
Yao
